src/cmds/qdel.c

```diff
@@ -58,6 +58,75 @@
 #include <pbs_config.h>   /* the master config generated by configure */
 #include <pbs_version.h>
 
+#define no_of_jobids 1000
+
+typedef struct {
+	int max_sz;
+	int job_ct;
+	char *server_name;
+	char **jobids;
+} pbs_svr_t;
+
+int no_of_pbsservers;
+
+void
+append_jobid(pbs_svr_t *svr, char *jobid)
+{
+	if (svr->max_sz == 0) {
+		svr->jobids = calloc(no_of_jobids + 1, PBS_MAXCLTJOBID + 1);
+		if (svr->jobids == NULL) {
+			fprintf(stderr, "qdel: unable to allocate memory.\n");
+			exit(2);
+		}
+		svr->max_sz = no_of_jobids;
+	} else if (svr->job_ct >= svr->max_sz) {
+		svr->max_sz *= 2;
+		svr->jobids = realloc(&svr->jobids, svr->max_sz);
+		if (svr->jobids == NULL) {
+			fprintf(stderr, "qdel: unable to allocate memory.\n");
+			exit(2);
+		}
+	}
+	
+	svr->jobids[svr->job_ct++] = jobid;
+}
+
+int
+add_jobid_to_pbs_servers(char *job_id, char *server_out, pbs_svr_t** pbs_servers)
+{
+	int i;
+	int found = 0;
+	
+tryagain:
+	for (i = 0; i < no_of_pbsservers; i++) {
+		if (pbs_servers[i]->server_name == NULL) {
+			pbs_servers[i]->server_name = strdup(server_out);
+			if (pbs_servers[i]->server_name == NULL) {
+				fprintf(stderr, "qdel: unable to allocate memory.\n");
+				exit(2);
+			}
+			append_jobid(pbs_servers[i], job_id);
```

Comment:
We cant use this function since these functions would allocate new heap memory and return the passed siring. Here the objective is we need to reuse the reallocated memory for *arv[].

Reply:
add_ptr_to_array does exactly that. it doesn't do strdup() like add_str_to_array

Reply:
please have a look into "add_ptr_to_array", it is counting the number of elements in each call and reallocating the size of the string array and string the new string. If we want to use this we need to refactor a lot of code in the scheduler wherever it is getting used. That refactor should be taken separately. I don't want to mix fixing that functions and extend this qdel refactoring work. 

If the function is a good performant to use, it makes sense to use it directly.

Reply:
We have using it in scheduler for long time @bremanandjk so I don't think it will cause any slowness. Also this code is client side so it won't matter much.

Reply:
I think we can't say that since we are using in the scheduler, we can use it here also. Even though it's a client program, if we have a huge number say(1 million), jobs we are going to iterate over all jobs in each job addition. like 1million factorial. Do you still think, it's not a performance issue.

Reply:
In our internal discussion, @nithinj we, decided to proceed with append_jobid().